src/hotspot/share/opto/node.cpp

```diff
@@ -1696,6 +1696,670 @@ bool Node::add_to_worklist(Node* n, Node_List* worklist, Arena* old_arena, Vecto
+      // all backward edges
+      Node* back = info->edge_bwd.at(i);
+      Info* back_info = find_info(back);
+      if (!back_info->is_marked()) {
+        // not yet found this on way back.
+        back_info->set_distance_from_target(info->distance_from_target() + 1);
+        if (back_info->distance_from_target() + back_info->distance() <= _max_distance) {
+          // total distance is small enough
+          back_info->set_mark();
+          backtrace.push(back);
+        }
+      }
+    }
+  }
+}
+
+void PrintBFS::select_shortest_path() {
+  Node* current = _target;
+  while (true) {
+    Info* info = find_info(current);
+    info->set_mark();
+    if (current == _start) {
+      break;
+    }
+    // first edge -> leads us one step closer to _start
+    current = info->edge_bwd.at(0);
+  }
+}
+
+int node_idx_cmp(Node** n1, Node** n2) {
+  return (*n1)->_idx - (*n2)->_idx;
+}
+
+// go through worklist in desired order, put the marked ones in print list
+void PrintBFS::sort() {
+  if (_traverse_inputs && !_traverse_outputs) {
+    // reverse order
+    for (int i = _worklist.size() - 1; i >= 0; i--) {
+      Node* n = _worklist.at(i);
+      Info* info = find_info(n);
+      if (info->is_marked()) {
+        _print_list.push(n);
+      }
+    }
+  } else {
+    // same order as worklist
+    for (uint i = 0; i < _worklist.size(); i++) {
+      Node* n = _worklist.at(i);
+      Info* info = find_info(n);
+      if (info->is_marked()) {
+        _print_list.push(n);
+      }
+    }
+  }
+  if (_sort_idx) {
+    _print_list.sort(node_idx_cmp);
+  }
+}
+
+// go through printlist and print
+void PrintBFS::print() {
+  if (_print_list.length() > 0 ) {
+    print_header();
+    for (int i = 0; i < _print_list.length(); i++) {
+      Node* n = _print_list.at(i);
+      print_node(n);
+    }
+  } else {
+    tty->print("No nodes to print.\n");
+  }
+}
+
+void PrintBFS::print_options_help(bool print_examples) {
+  tty->print("Usage: node->dump_bfs(int max_distance, Node* target, char* options)\n");
+  tty->print("\n");
+  tty->print("Use cases:\n");
+  tty->print("  BFS traversal: no target required\n");
+  tty->print("  shortest path: set target\n");
+  tty->print("  all paths: set target and put 'A' in options\n");
+  tty->print("  detect loop: subcase of all paths, have start==target\n");
+  tty->print("\n");
+  tty->print("Arguments:\n");
+  tty->print("  this/start: staring point of BFS\n");
+  tty->print("  target:\n");
+  tty->print("    if nullptr: simple BFS\n");
+  tty->print("    else: shortest path or all paths between this/start and target\n");
+  tty->print("  options:\n");
+  tty->print("    if nullptr: same as \"cdmxo@B\"\n");
+  tty->print("    else: use combination of following characters\n");
+  tty->print("      h: display this help info\n");
+  tty->print("      H: display this help info, with examples\n");
+  tty->print("      +: traverse in-edges (on if neither + nor -)\n");
+  tty->print("      -: traverse out-edges\n");
```

Comment:
I think I will keep it, as `dump` also by default only traverses in-edges.
Additionally, it can be quite confusing to traverse input and output edges at the same time.